### PR TITLE
[Dynamic Protocol State] KV store state machine

### PR DESCRIPTION
**Pull request by durkmurder**
https://github.com/onflow/flow-go/issues/5312

### Context

This PR implements a dedicated state machine in context of Dynamic Protocol State to process service events that update the KV store. 

This PR relies on previous PR created by Jordan that introduces new service event type for upgrading protocol state version. 
- I've added processing of the upgrade event by the state machine. 
- Additionally I had to extend model interface to return `VersionUpgrade` and be able to set it. 
- All models MUST support this notion to have the KV store upgradable. 

I've took the liberty to update package structure to avoid nasty import cycles and to increase modularity in overall. 

:warning: 
I've made a debatable choice of using `EpochCommitSafetyThreshold` as threshold for activation of upgrade event. I feel this is legit since it serves the same purpose but maybe we should change the naming then. Happy to use a separate param as well.

### CODE REVIEW

File: state/protocol/protocol_state/kvstore/models.go [hunk 1]
```diff
@@ -13,10 +30,19 @@ package kvstore
...
 
-var _ Reader = new(modelv0)
-var _ API = new(modelv0)
+var _ protocol_state.Reader = new(modelv0)
+var _ protocol_state.API = new(modelv0)
+
+func (model *modelv0) ID() flow.Identifier { return flow.MakeID(model) }
+
+func (model *modelv0) Clone() protocol_state.API {
+	cpy := *model
```
Comment (jordanschalm) [inline state/protocol/protocol_state/kvstore/models.go:43]: ⚠️  `VersionUpgrade` is a pointer field of `modelv0`, so this isn't really a deep copy. We should copy pointer fields as well.

File: state/protocol/protocol_state/kvstore/models.go [hunk 2]
```diff
@@ -53,8 +81,15 @@ type modelv1 struct {
...
 }
 
-var _ Reader = new(modelv1)
-var _ API = new(modelv1)
+var _ protocol_state.Reader = new(modelv1)
+var _ protocol_state.API = new(modelv1)
+
+func (model *modelv1) ID() flow.Identifier { return flow.MakeID(model) }
+
+func (model *modelv1) Clone() protocol_state.API {
```
Comment (jordanschalm) [inline state/protocol/protocol_state/kvstore/models.go:89]: same comment here, this isn't a deep copy

File: state/protocol/protocol_state/kvstore/statemachine.go [hunk 3]
```diff
@@ -0,0 +1,104 @@
...
+	"github.com/onflow/flow-go/state/protocol/protocol_state"
+)
+
+// ProcessingStateMachine is a dedicated structure that encapsulates all logic for evolving KV store, based on the content
+// of a new block.
+// ProcessingStateMachine processes a subset of service events that are relevant for the KV store, and ignores all other events.
+// Each relevant event is validated before it is applied to the KV store.
+// All updates are applied to a copy of parent KV store, so parent KV store is not modified.
+// A separate instance should be created for each block to process the updates therein.
+type ProcessingStateMachine struct {
```
Comment (jordanschalm) [inline state/protocol/protocol_state/kvstore/statemachine.go:17]:
```suggestion
type StateMachine struct {
```
Would suggest to just call this `kvstore.StateMachine`, as it is the only state machine in this package.

File: state/protocol/protocol_state/kvstore/statemachine.go [hunk 4]
```diff
@@ -0,0 +1,104 @@
...
+		// To switch the protocol version, replica needs to process a block with a view >= activation view.
+		// But we cannot activate a new version till the block containing the seal is finalized because we cannot switch between chain forks.
+		// The problem is that finality is local to each node due to the nature of the consensus algorithm itself.
+		// We would like to guarantee that all nodes switch the protocol version at exactly the same block.
+		// To guarantee that all nodes switch the protocol version at exactly the same block, we require that the
+		// activation view is higher than the sealing view + Δ when accepting the event. Δ represents the finalization lag
+		// to give time for replicas to finalize the block containing the seal for the version upgrade event.
+		// When replica reaches activation view and the latest finalized protocol state knows about the version upgrade,
+		// then it's safe to switch the protocol version.
+		if m.view+m.params.EpochCommitSafetyThreshold() >= versionUpgrade.ActiveView {
```
Comment (jordanschalm) [inline state/protocol/protocol_state/kvstore/statemachine.go:72]: I agree with using this existing parameter, as the underlying assumption is the same. I suggest we change the name to something like `FinalizationSafetyThreshold`.

File: state/protocol/protocol_state/kvstore/statemachine.go [hunk 5]
```diff
@@ -0,0 +1,104 @@
...
+			return protocol.NewInvalidServiceEventErrorf("invalid protocol state version upgrade view %d -> %d: %w",
```
Comment (jordanschalm) [inline state/protocol/protocol_state/kvstore/statemachine.go:73]: We should ignore this kind of service event (caller should interpret `InvalidServiceEventError` as expected, maybe log a warning, but otherwise proceed along the happy path. 

At the moment I think that is true for _all_ `InvalidServiceEventError`s from the KVStore, but that may not always be the case. If we have KVStore-modifying service events for which failures are not expected, we will need to make the sentinels more granular. 

No action needed here, just making a note.

File: state/protocol/protocol_state/kvstore/statemachine.go [hunk 6]
```diff
@@ -0,0 +1,104 @@
...
+				m.parentState.GetProtocolStateVersion(), versionUpgrade.NewProtocolStateVersion, ErrInvalidUpgradeVersion)
+		}
+
+		activator := &protocol_state.ViewBasedActivator[uint64]{
+			Data:           versionUpgrade.NewProtocolStateVersion,
+			ActivationView: versionUpgrade.ActiveView,
+		}
+		m.state.SetVersionUpgrade(activator)
+
+	default:
```
Comment (jordanschalm) [inline state/protocol/protocol_state/kvstore/statemachine.go:88]:
```suggestion
	// Service events not explicitly expected are ignored
	default:
```

**Review by jordanschalm — COMMENTED** [review]
Nice work. 
- I think we use `EpochCommitSafetyThreshold`, but rename it. Renaming could be a separate PR, as that is referenced in several places, including CLI utils
- Otherwise largely documentation-related and non-functional suggestions

**Review by jordanschalm — APPROVED** [review]

**Comment by AlexHentschel** [comment]
wanted to comment on your statement in the PR description:
> ⚠️
> I've made a debatable choice of using EpochCommitSafetyThreshold as threshold for activation of upgrade event. I feel this is legit since it serves the same purpose but maybe we should change the naming then. Happy to use a separate param as well.

I think that is a fine choice. The default value is 1000 which is about 20mins, which is already really short for a software upgrade to be widely deployed. There might be edge cases where we would desire a faster activation but it is hard for me to come up with meaningful scenarios. 

TLDR: 👍 on using he same parameter

File: state/protocol/protocol_state/state/mutator.go [hunk 7]
```diff
@@ -33,8 +34,8 @@ type stateMutator struct {
 	results                          storage.ExecutionResults
 	setups                           storage.EpochSetups
 	commits                          storage.EpochCommits
-	stateMachine                     ProtocolStateMachine
-	epochFallbackStateMachineFactory func() (ProtocolStateMachine, error)
+	stateMachine                     epochs.ProtocolStateMachine
+	epochFallbackStateMachineFactory func() (epochs.ProtocolStateMachine, error)
```
[the remark this answers is not in the document]
Comment (durkmurder) [reply state/protocol/protocol_state/state/mutator.go:38]: I was planning to wire up it in separate issue, possibly in https://github.com/onflow/flow-go/issues/5319. 
I tend to put it alongside with `stateMachine` in `stateMutator` and filter events by type to decide who should handle which event. The idea was that `stateMutator` knows about different state machines and routes events to the target state machine.

File: state/protocol/protocol_state/kvstore/models.go (hunk 1 above)
Comment (durkmurder) [reply state/protocol/protocol_state/kvstore/models.go]: Yeap, good catch, I've added `VersionUpgrade` after adding `Clone`. I have used a 3rd party library which performs a generic deep copy and added tests: [`2416aa1` (#5513)](https://github.com/onflow/flow-go/pull/5513/commits/2416aa1d03cc8b45dd8c23e66fe31b20d62a6896)

File: state/protocol/protocol_state/kvstore/statemachine.go (hunk 5 above)
Comment (durkmurder) [reply state/protocol/protocol_state/kvstore/statemachine.go:73]: Specifically for that I am returning a sentinel so caller can distinguish between cases and ignore needed event.

File: state/protocol/protocol_state/kvstore.go [hunk 8]
```diff
@@ -0,0 +1,108 @@
...
+	SetInvalidEpochTransitionAttempted(attempted bool) error
+}
+
+// KeyValueStoreStateMachine implements a low-level interface for state-changing operations on the key-value store.
+// It is used by higher level logic to evolve the protocol state when certain events that are stored in blocks are observed.
+// The KeyValueStoreStateMachine is stateful and internally tracks the current state of key-value store.
+// A separate instance is created for each block that is being processed.
+type KeyValueStoreStateMachine interface {
+	// Build returns updated key-value store model, state ID and a flag indicating if there were any changes.
+	Build() (updatedState KVStoreReader, stateID flow.Identifier, hasChanges bool)
```
Comment (AlexHentschel) [inline state/protocol/protocol_state/kvstore.go:93]: In my mind, it would be great if the `Build` method returned a representation of the protocol state snapshot that permits mutations. So specifically `KVStoreAPI`. For instance, if we cache the results to apply the state-changing events from the next block (one we receive it), we would want them represented via the mutable `KVStoreAPI`  (hope that makes sense).

```suggestion
	Build() (updatedState KVStoreAPI, snapshotID flow.Identifier, hasChanges bool)
```

File: state/protocol/protocol_state/kvstore/errors.go
Comment (AlexHentschel) [inline state/protocol/protocol_state/kvstore/errors.go:1]: The colon (`:`) is used as a separator by wrapped error messages. It might be confusing for an engineer who sees an error message in the logs to find the respective source code if we re-use the colon within the error messages.

I think it would be easy to fix:
* https://github.com/onflow/flow-go/blob/2416aa1d03cc8b45dd8c23e66fe31b20d62a6896/state/protocol/protocol_state/kvstore/errors.go#L10 you could use the error message "no value for requested key in Protocol State's kvstore"
* https://github.com/onflow/flow-go/blob/2416aa1d03cc8b45dd8c23e66fe31b20d62a6896/state/protocol/protocol_state/kvstore/errors.go#L17 you could use the error message "Protocol State's kvstore does not support the specified key at this version"
* https://github.com/onflow/flow-go/blob/2416aa1d03cc8b45dd8c23e66fe31b20d62a6896/state/protocol/protocol_state/kvstore/errors.go#L23  you could use the error message "unsupported version for the Protocol State's kvstore"
* https://github.com/onflow/flow-go/blob/2416aa1d03cc8b45dd8c23e66fe31b20d62a6896/state/protocol/protocol_state/kvstore/errors.go#L29 you could use the error message "invalid upgrade version for the Protocol State's kvstore"
* https://github.com/onflow/flow-go/blob/2416aa1d03cc8b45dd8c23e66fe31b20d62a6896/state/protocol/protocol_state/kvstore/errors.go#L33  you could use the error message "invalid activation view for the new Protocol State version"

**Review by AlexHentschel — COMMENTED** [review]
Ver nice work. Mostly comments regarding documentation. My only design concern is regarding the `KVStoreAPI`'s `Clone` method. Added a detailed comment and suggestion [here]().

File: state/protocol/protocol_state/kvstore.go [hunk 9]
```diff
@@ -0,0 +1,108 @@
...
+
+	// GetProtocolStateVersion returns the current Protocol State version.
+	// The Protocol State version specifies the current version of the Protocol State,
+	// including the key-value store. Changes in the protocol state version
+	// correspond to changes in the set of key-value pairs which are supported,
+	// and which model is used for serialization.
+	// It can be updated by an UpdateKVStoreVersion service event.
+	GetProtocolStateVersion() uint64
+
+	// GetVersionUpgrade returns the upgrade version of protocol.
```
[the remark this answers is not in the document]
Comment (durkmurder) [reply state/protocol/protocol_state/kvstore.go:48]: Well, `GetVersionUpgrade`stores the activation view and version that can be used for next upgrade. It is NOT the version that was used to create underlying structure that implements API, for that we need to use `GetProtocolStateVersion`

File: state/protocol/protocol_state/kvstore.go [hunk 10]
```diff
@@ -0,0 +1,108 @@
...
+
+// KVStoreAPI is the latest read-writer interface to the Protocol State key-value store.
+// It is backward-compatible with all versioned model types defined in the models.go
+// for this software version.
+type KVStoreAPI interface {
+	KVStoreReader
+
+	// Clone returns a deep copy of the KVStoreAPI.
+	// This is used to create a new instance of the KVStoreAPI to avoid mutating the original.
+	Clone() KVStoreAPI
```
[the remark this answers is not in the document]
Comment (durkmurder) [reply state/protocol/protocol_state/kvstore.go:71]: Thanks for detailed explanation and argumentation, the more I read it, the more I tend to get rid of `Clone/Replicate` as I have implemented it in the first place. I feel like we are mixing two concepts, _mutating_ data and _creation_ process. 
I have tried to replicate the same structure as we have with epoch protocol state but it doesn't seem to work out. Meaning the state machine doesn't have enough context to provide a correct version to make a replication with correct version even if it had we are mixing creation pattern and behavior(factory + state machine). 

In my mind it works well where the `stateMutator` acts as a factory and creates a correct versioned `KVStoreMutator` which is then passed to the state machine. Similarly to what we have with the epoch state machine. This way we separate creation and mutation. This means we don't need `Replicate` or `Clone` as part of API and the state machine won't take a copy on it's own, instead it will rely on the instance that will be received via constructor from a component that has needed context to create it. 

Lets continue this discussion.

File: state/protocol/protocol_state/kvstore.go (hunk 9 above)
[the remark this answers is not in the document]
Comment (AlexHentschel) [reply state/protocol/protocol_state/kvstore.go:48]: thanks for the explanation. I would suggest to update the documentation of `GetProtocolStateVersion` to 

```golang
// GetProtocolStateVersion returns the Protocol State Version that created the specific
// Snapshot backing this interface instance. Slightly simplified, the Protocol State
// Version defines the key-value store's data model (specifically, the set of all keys
// and the respective type for each corresponding value). 
// Generally, changes in the protocol state version correspond to changes in the set
// of key-value pairs which are supported, and which model is used for serialization.
// The protocol state version is updated by UpdateKVStoreVersion service events.
GetProtocolStateVersion() uint64
```

I have tried to avoid using "current" and instead talked about the  "Protocol State Version that created the specific
Snapshot backing this interface instance."

File: state/protocol/protocol_state/kvstore.go (hunk 10 above)
[the remark this answers is not in the document]
Comment (AlexHentschel) [reply state/protocol/protocol_state/kvstore.go:65]: ### Summary of sync (Yurii, Jordan, Alex, March 12, 2024)

1. `Replicate(..)` or `Clone` must produce valid Snapshot, because in many blocks there are no service events
2. from 1. it follows that we don't need builder pattern 
3. extending `stateMutator` to be the orchestrator of all state machines (currently only orchestrates Epoch State machine) 
    * `stateMutator` will call `Replicate`
    * state machines are applied, they mutate the combined state in-place 
    * we `Build` method in  `stateMutator`  where we could add sanity checks

We want to move the entire Epoch data into the KV-store, but that is beyond the current deliverables.

File: state/protocol/protocol_state/kvstore/models.go [hunk 11]
```diff
@@ -34,14 +45,33 @@ type modelv0 struct {
...
 
-func (model *modelv0) Clone() protocol_state.API {
-	cpy := *model
-	return &cpy
+// Replicate instantiates a Protocol State Snapshot of the given `protocolVersion`.
+// It clones existing snapshot if `protocolVersion = 0` and performs a migration if `protocolVersion = 1`.
+// Expected errors during normal operations:
+//   - ErrIncompatibleVersionChange if replicating the Parent Snapshot into a Snapshot
+//     with the specified `protocolVersion` is not supported.
+func (model *modelv0) Replicate(protocolVersion uint64) (protocol_state.KVStoreMutator, error) {
```
Comment (jordanschalm) [inline state/protocol/protocol_state/kvstore/models.go:59]: We discussed having `Replicate` live alongside the models and be an "atomic" process, in the sense that we will strictly return a valid KVStore model from `Replicate`. 

In my mind that constraint includes making sure that a KVStore which has been upgraded to protocol version N does not include a pending `VersionUpgrade` for protocol version N. However we're not doing that here.

I'm wondering if `Replicate` should instead receive as input the **block view**, and internally decide whether an upgrade is needed, based on the `VersionUpgrade` field? 

That way, the higher-level logic would not need to deal with:
- deciding when the version becomes active
- removing the old `VersionUpgrade` following the upgrade

Based on [your comment here](https://github.com/onflow/flow-go/pull/5513/files/856299a96640f57110e2a792cf59bc9bcc9a6f9d..ba1f763f08d56fff590b4c9c3e70399c72749f45#diff-d5d9a8e1e06fcaf771c56b400d81c56429405675f36121b42294c3cb3eee38c0R90), I assume your plan is to keep outdated `VersionUpgrade`s in the protocol state until they are replaced. If we choose that route, this comment is mostly irrelevant (we could still choose to pass in view rather than `protocolVersion` to `Replicate`).

Overall, I'm happy to merge this as-is and make a final decision when we hook up the state machine.

**Review by jordanschalm — APPROVED** [review]
Reviewed the diff since our last discussion. Looks good overall -- I have added a suggestion for how to organize the upgrade logic (in `Replicate` vs in `Replicate`s caller), although I am happy to punt that discussion to the PR where we hook up the state machine.

File: state/protocol/protocol_state/kvstore/models.go (hunk 11 above)
Comment (durkmurder) [reply state/protocol/protocol_state/kvstore/models.go:59]: I was thinking about cleaning `VersionUpgrade` when we replicate, but the process still waits for trigger. Happy to discuss after. For now leaving as it is.

File: state/protocol/protocol_state/epochs/protocol_statemachine.go [hunk 12]
```diff
@@ -26,10 +26,10 @@ type protocolStateMachine struct {
 
 var _ ProtocolStateMachine = (*protocolStateMachine)(nil)
 
-// newStateMachine creates a new protocol state protocolStateMachine.
+// NewStateMachine creates a new protocol state protocolStateMachine.
```
[the remark this answers is not in the document]
Comment (AlexHentschel) [reply state/protocol/protocol_state/epochs/protocol_statemachine.go:30]: I would strongly advocate for `parallel` state machines. If we want to avoid getting ourselves in a hell of logical interdependencies, I think there is only one self-consistent approach:
* Each state machine has its own portion of the state that it _exclusively_ mutates. 
* Therefore, the overall state is a product state of all the state-machine specific sub-states
* The overall state machine is formalized as `Orthogonal Regions` [[1](https://en.wikipedia.org/wiki/UML_state_machine#Orthogonal_regions), [2](https://www.state-machine.com/doc/Pattern_Orthogonal.pdf)] composed of the individual state machines. 

For an intuition, this is often framed as "concurrent state machines" (or product-state-machines). 

Note: 
* This is an explicit limitation, that we purposefully impose on the Protocol State's KV-store. Mathematically, we say that the overall state machine $\mathcal{S}$ must be decomposable into $\mathcal{S} = \mathcal{S}_0 \otimes \mathcal{S}_1 \otimes \ldots  \otimes \mathcal{S}_j$
* This does not prevent us of from implementing sequential state machines for certain events. It is just that we would need to bundle these sequential state machines into a composite state machine (conceptually a processing pipeline). Something like saying that $\mathcal{S}_i$ (for  some specific $i \in [0,\ldots,j]$) can be decomposed into  $\mathcal{S}_i = \mathcal{S}^{(\kappa)}_i \cdots \mathcal{S}^{(1)}_i\mathcal{S}^{(0)}_i$

I hope the mathematical notation helps to explain 😅 ($\otimes$ is not quite a [tensor product](https://en.wikipedia.org/wiki/Tensor_product) because we are not operating on vector spaces, but conceptually it is the same thing, product spaces and product operators)

[the remark this answers is not in the document]
Comment (AlexHentschel) [reply state/protocol/protocol_state/epochs/protocol_statemachine.go:30]: I think what we are running into here in this particular case is the lack of separation between:
* the framework that evolves the protocol state in my mind is [`MutableProtocolState`](https://github.com/onflow/flow-go/blob/ba1f763f08d56fff590b4c9c3e70399c72749f45/state/protocol/protocol_state/state/protocol_state.go#L83) and [`stateMutator`](https://github.com/onflow/flow-go/blob/ba1f763f08d56fff590b4c9c3e70399c72749f45/state/protocol/protocol_state/state/mutator.go#L47)

In my opinion, everything related to the Epoch should go be governed by one state machine from the perspective of  the `stateMutator`:
* Internally, the epoch state machine would then need to handle the fallback.  So we'll have to do some more work to modularize the logic (including [this code for](https://github.com/onflow/flow-go/blob/ba1f763f08d56fff590b4c9c3e70399c72749f45/state/protocol/protocol_state/state/mutator.go#L205-L258) example), because we shouldn't just error back to the `stateMutator` and re-process all events.  
* I think this logic of mediating between happy path Epoch and fallback should go into the top-level `epoch.stateMachine` 

I think this top-level `epoch.stateMachine` would be the place to document the concurrent processing of the Protocol State Machines. We should emphasize that the `epoch.stateMachine`  is then one of the concurrent state machines running 
Nevertheless, I think the **main place** where we should **document the state machine processing model** is the [`StateMutator` interface](https://github.com/onflow/flow-go/blob/00e4924e0d1b2d69223e5db722b8e6f15ce92399/state/protocol/protocol_state.go#L96-L109) and [`stateMutator` implementation](https://github.com/onflow/flow-go/blob/ba1f763f08d56fff590b4c9c3e70399c72749f45/state/protocol/protocol_state/state/mutator.go#L19-L32).

BUT, lets do that in a follow-up PR. If we agree on this route, may I ask for an issue to be created please 🙏 🙇

File: state/protocol/protocol_state/kvstore/statemachine.go (hunk 5 above)
Comment (AlexHentschel) [reply state/protocol/protocol_state/kvstore/statemachine.go:73]: similarly to [my comment below](https://github.com/onflow/flow-go/pull/5513#discussion_r1529510443), I think we are lacking the necessary modularization in the code here to adequately reflect our architectural approach. 

I am going to use the same notation here as in [my comment below](https://github.com/onflow/flow-go/pull/5513#discussion_r1529468350). 
* At the moment, we have the epoch state machine logic. In my comment below, I am [suggesting](https://github.com/onflow/flow-go/pull/5513#discussion_r1529510443) to consolidate this logic all in the [`epochs` package](https://github.com/onflow/flow-go/tree/ba1f763f08d56fff590b4c9c3e70399c72749f45/state/protocol/protocol_state/epochs), with a new `EpochStateMachine` being the top level logic. 

  Lets denote the `EpochStateMachine` as $\mathcal{S}_0$
* In the `kvstore` package, there is conceptually a separate state machine that consumes the `ProtocolStateVersionUpgrade`. Specifically, I think that [this logic](https://github.com/onflow/flow-go/blob/ba1f763f08d56fff590b4c9c3e70399c72749f45/state/protocol/protocol_state/kvstore/statemachine.go#L58-L119) should be represented as its own state machine, for example `PSVersionUpgradeStateMachine`

  Lets denote the `PSVersionUpgradeStateMachine` as $\mathcal{S}_1$

* Furthermore, lets consider a future scenario where we want to add a separate Version Upgrade Mechanism for ENs+VNs (similar to the existing HCU). That also utilizes the protocol state's KV store, with its own separate key. 
   * I would suggest that this logic would be represented as yet another orthogonal state machine, lets call it `ExecutionVersionUpgradeStateMachine`
   * the `ExecutionVersionUpgradeStateMachine` uses the KV store, so we put it in the `kvstore` package
   
  Lets denote the `ExecutionVersionUpgradeStateMachine` as $\mathcal{S}_2$

So in total, we have the orthogonal state machines  $\mathcal{S}_0$ (for epochs), $\mathcal{S}_1$ (for updating the protocol state) and $\mathcal{S}_2$ (for EN HCUs).

The composite Protocol State Machine (largely [this logic](https://github.com/onflow/flow-go/blob/ba1f763f08d56fff590b4c9c3e70399c72749f45/state/protocol/protocol_state/state/mutator.go#L168-L215) in `stateMutator` atm) would then operates directly on  $\mathcal{S}_0, \mathcal{S}_1, \mathcal{S}_2$ 
  * `stateMutator` would essentially implement $\mathcal{S} = \mathcal{S}_0 \otimes \mathcal{S}_1  \otimes \mathcal{S}_2$, where $\mathcal{S}_0, \mathcal{S}_1, \mathcal{S}_2$ are initialized when `stateMutator` is constructed
  * `stateMutator` would then take the list of service event in a block and multiplex them to each `\mathcal{S}_0, \mathcal{S}_1, \mathcal{S}_2`. In other words, each state machine receives all the events and just picks the events that are relevant for it. 
  * Not sure, but it might help code clarity, if we change the computation model a bit, where state machine $\mathcal{S}_i$ receives as as input: 
     * the parent state, 
     * the replicated parent state (potential version upgrade already applied) to write its outputs into
     * and a stream of service events

Comment (AlexHentschel) [reply state/protocol/protocol_state/kvstore/statemachine.go:73]: beyond this PR, but wanted to document my throughs:

```golang
// OrthogonalStoreStateMachine represents a state machine, that exclusively evolves its state P.
// The state's specific type P is kept as a generic. Generally, P is the type corresponding
// to one specific key in the Key-Value store.
//
// The Flow protocol defines its protocol state 𝓅 as the composition of disjoint sub-states P0, P1, ..., Pj
// Formally, we write 𝒫 = P0 ⊗ P1 ⊗  ... ⊗ Pj, where '⊗' denotes the product state. We loosely associate
// each P0, P1 ... with one specific key-value entry in the store. Correspondingly, we have one conceptually
// independent state machine S0, S1, ... operating each on their own respective element P0, P1, ...
// A one-to-one correspondence between kv-entry and state machine should be the default, but is not
// strictly required. However, the strong requirement is that no entry is operated on my more than one
// state machine.
//
// Orthogonal State Machines:
// Orthogonality means that state machines can operate completely independently and work on disjoint
// sub-states. They all consume the same inputs: the ordered sequence of Service Events sealed in one block.
// In other words, each state machine S0, S1, ... has full visibility, but each draws their on independent
// conclusions (maintain their own exclusive state).
// This is a deliberate design choice. Thereby the default is favouring modularity and strong logical
// independence. This is very beneficial for managing complexity in the long term.
// We emphasize that this architecture choice does not prevent us of from implementing sequential state
// machines for certain events. It is just that we would need to bundle these sequential state machines
// into a composite state machine (conceptually a processing pipeline).
//
// Formally we write:
//   - The overall protocol state 𝒫 is composed of disjoint substates 𝒫 = P0 ⊗ P1 ⊗  ... ⊗ Pj
//   - For each state Pi, we have a dedicated state machine Si that exclusively operates on Pi
//   - The state machines can be formalized as orthogonal regions of the composite state machine
//     𝒮 = S0 ⊗ S1 ⊗  ... ⊗ Sj. (Technically, we represent the state machine by its state-transition
//     function. All other details of the state machine are implicit.)
//   - The state machine 𝒮 being in state 𝒫 and observing the input ξ = x0·x1·x2·..·xz will output
//     state 𝒫'. To emphasize that a certain state machine 𝒮 exclusively operates on state 𝒫, we write
//     𝒮[𝒫] = S0[P0] ⊗ S1[P1] ⊗ ... ⊗ Sj[Pj]
//     Observing the events ξ the output state 𝒫' is
//     𝒫' = 𝒮[𝒫](ξ) = S0[P0](ξ) ⊗ S1[P1](ξ) ⊗ ... ⊗ Sj[Pj](ξ) = P'0 ⊗ P'1 ⊗  ... ⊗ P'j
//     Where each state machine Si individually generated the output state Si[Pi](ξ) = P'i
//
// The Protocol State is the framework, which orchestrates the orthogonal state machines,
// feeds them with inputs, post-processes the outputs and overall manages state machines' life-cycle
// from block to block. New key-value pairs and corresponding state machines can easily be added
// by implementing the following interface (state machine) and adding a new entry to the KV store.
type OrthogonalStoreStateMachine[P any] interface {

	// Build returns:
	//   - hasChanges: flag whether there were any changes; otherwise, `updatedState` and `stateID` equal the parent state
	//   - updatedState: the ProtocolState after applying all updates.
	//   - stateID: the hash commitment to the `updatedState`
	//   - dbUpdates: database updates necessary for persisting the updated protocol state's *dependencies*.
	//     If hasChanges is false, updatedState is empty. Caution: persisting the `updatedState` itself and adding
	//     it to the relevant indices is _not_ in `dbUpdates`. Persisting and indexing `updatedState` is the responsibility
	//     of the calling code (specifically `FollowerState`).
	//
	// updated protocol state entry, state ID and a flag indicating if there were any changes.
	Build() (hasChanges bool, updatedState P, dbUpdates []transaction.DeferredDBUpdate)

	// Build returns updated key-value store model, state ID and a flag indicating if there were any changes.
	//Build() (updatedState KVStoreReader, stateID flow.Identifier, hasChanges bool)

	// ProcessUpdate updates the current state of key-value store.
	// KeyValueStoreStateMachine captures only a subset of all service events, those that are relevant for the KV store. All other events are ignored.
	// Implementors MUST ensure KeyValueStoreStateMachine is left in functional state if an invalid service event has been supplied.
	// Expected errors indicating that we have observed an invalid service event from protocol's point of view.
	// 	- `protocol.InvalidServiceEventError` - if the service event is invalid for the current protocol state.
	ProcessUpdate(update *flow.ServiceEvent) error

	// View returns the view that is associated with this KeyValueStoreStateMachine.
	// The view of the KeyValueStoreStateMachine equals the view of the block carrying the respective updates.
	View() uint64

	// ParentState returns parent state that is associated with this state machine.
	ParentState() KVStoreReader
}
```

**Review by AlexHentschel — APPROVED** [review]
My comments are on the architecture level. I think this PR is good to go. I would like to iterate on the design to get to more modularity and a clearer representation of the conceptual aspects. I felt a variety of detailed comments (partially on existing discussions).

If we could get this issued up, that would be great. But before we write issues, let's make sure we are on the same page regarding the design outlook.
